elector/elector.go

```diff
@@ -0,0 +1,286 @@
+// it's znode data.  This will typically be the IP address of the client of
+// the elector.
+//
+// The 'basePath' parameter is the znode under which the leader election will
+// happen.
+//
+// The 'acl' will be set on any nodes that must be created
+func Start(ident string, basePath string, acl []zk.ACL, connector Connector) (*Elector, error) {
+	if strings.TrimSpace(ident) == "" {
+		return nil, errors.New("ident must not be blank")
+	}
+	if acl == nil {
+		acl = zk.WorldACL(zk.PermAll)
+	}
+	conn, zkEvents, err := connector.Connect()
+	if err != nil {
+		return nil, err
+	}
+	elector := &Elector{
+		acl:      acl,
+		ident:    ident,
+		conn:     conn,
+		basePath: basePath,
+		events:   make(chan Event),
+		closer:   connector.Close,
+	}
+	if err := elector.initialize(); err != nil {
+		return nil, errors.Wrap(err, "elector initialization failed")
+	}
+	go elector.start(zkEvents)
+	return elector, nil
+}
+
+// LeaderIdent returns the current leader, or "" if no current leader is
+// known yet.
+func (e *Elector) LeaderIdent() string {
+	e.mut.Lock()
+	defer e.mut.Unlock()
+	return e.leaderIdent
+}
+
+// Events returns a channel on which Events will be sent.
+func (e *Elector) Events() <-chan Event {
+	return e.events
+}
+
+// Close closes the underlying ZK connection. Clients should call Close() when
+// abandoning elector efforts in order to quickly delete any ephemeral nodes
+// that were created as a part of the election process.
+func (e *Elector) Close() error {
+	return e.closer()
+}
+
+// initialize sets up the basePath if necessary
+func (e *Elector) initialize() error {
+	exists, _, err := e.conn.Exists(e.basePath)
+	if err != nil {
+		return errors.Wrapf(err, "could not check if base path %s exists", e.basePath)
+	}
+	if exists {
+		return nil
+	}
+	segments := strings.Split(e.basePath, "/")
+	create := "/"
+	for _, segment := range segments {
+		create = path.Join(create, segment)
+		exists, _, err := e.conn.Exists(create)
+		if err != nil {
+			return errors.Wrapf(err, "could not check path '%s'", create)
+		}
+		if exists {
+			continue
+		}
+		_, err = e.conn.Create(create, []byte{}, 0, e.acl)
+		if err != nil {
+			return errors.Wrapf(err, "could not create path '%s'", create)
+		}
+	}
+	return nil
+}
+
+func (e *Elector) start(zkEvents <-chan zk.Event) {
+	err := func() error {
+		lockPath, err := e.conn.CreateProtectedEphemeralSequential(
+			e.basePath+"/lock-",
+			[]byte(e.ident),
+			e.acl)
+		if err != nil {
+			return errors.Wrap(err, "could not create lock node")
+		}
+
+		firstLeaderUpdate := true
+		updateFunc := func(children []string) error {
+			isLeader, leaderNode, err := determineLeader(lockPath, children)
+			if err != nil {
+				return errors.Wrap(err, "could not determine leader")
+			}
+			leaderIdent, err := e.getIdentFromNode(leaderNode)
+			if err != nil {
+				return errors.Wrap(err, "could not get ident from node")
+			}
+			e.setLeaderIdent(leaderIdent)
```

Comment:
it's unclear to me why these two functions `setLeaderIdent` and `sendLeader` are split up: there's potential for 3 mutex lock/unlock cycles across both calls, which seems strange. possible to consolidate the state changes to a single func that w/ a single mutex lock/unlock cycle?

Reply:
Thanks. yeah, that's unnecessarily complex.